core/store/src/flat_state.rs

```diff
@@ -405,9 +396,41 @@ impl FlatStateDelta {
     pub fn apply_to_flat_state(self, _store_update: &mut StoreUpdate) {}
 }
 
+/// `FlatStateDelta` which uses hash of raw `TrieKey`s instead of keys themselves.
+/// Used to reduce memory used by deltas and serves read queries.
+pub struct CachedFlatStateDelta(HashMap<CryptoHash, Option<ValueRef>>);
+
+impl From<FlatStateDelta> for CachedFlatStateDelta {
+    fn from(delta: FlatStateDelta) -> Self {
+        Self(delta.0.iter().map(|(key, value)| (hash(key), value.clone())).collect())
+    }
+}
+
+impl CachedFlatStateDelta {
+    /// Assumed size in bytes for the cache entry. Estimated as 69 bytes for `CryptoHash` and `Option<ValueRef>` +
+    /// 70% hashmap overhead per entry ~= 117 bytes.
+    const ENTRY_SIZE: u64 = 117;
+
+    /// Returns `Some(Option<ValueRef>)` from delta for the given key. If key is not present, returns None.
+    pub fn get(&self, key: &[u8]) -> Option<Option<ValueRef>> {
+        self.0.get(&hash(key)).cloned()
+    }
+
+    /// Returns number of all entries.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Size of all entries. May be changed if we implement inlining of `ValueRef`s.
+    fn total_size(&self) -> u64 {
+        (self.0.len() as u64) * Self::ENTRY_SIZE
```

Comment:
Let's try to be a bit more precise here. std `HashMap` is based on [Swiss Table](https://abseil.io/blog/20180927-swisstables). In my understanding it stores all the data in flat array + a couple of bits of "control data" per entry (that is how I understood it from [the talk](https://www.youtube.com/watch?v=ncHmEUmJZf4&t=3s)). So in our case I suggest using `self.0.capacity() * (sizeof<Key> + sizeof<Value>)` and ignore control data overhead to keep it simple.